src/Metalama.Patterns.NotifyPropertyChanged/Implementation/Natural/NaturalAspect.cs

```diff
@@ -0,0 +1,392 @@
+    {
+        // NB: The selection logic here must be kept in sync with the logic in the OnUnmonitoredObservablePropertyChanged template.
+
+        ctx.PropertyPathsForOnChildPropertyChangedMethodAttribute.AddRange(
+            ctx.DependencyGraph.DescendantsDepthFirst()
+                .Where(
+                    n => n.InpcBaseHandling switch
+                    {
+                        InpcBaseHandling.OnUnmonitoredObservablePropertyChanged when ctx.OnUnmonitoredObservablePropertyChangedMethod.WillBeDefined => true,
+                        InpcBaseHandling.OnPropertyChanged when n.HasChildren => true,
+                        _ => false
+                    } )
+                .Select( n => n.DottedPropertyPath ) );
+    }
+
+    private static void IntroduceOnPropertyChangedMethod( BuildAspectContext ctx )
+    {
+        var isOverride = ctx.BaseOnPropertyChangedMethod != null;
+
+        var result = ctx.Builder.Advice.IntroduceMethod(
+            ctx.Target,
+            nameof(OnPropertyChanged),
+            IntroductionScope.Instance,
+            isOverride ? OverrideStrategy.Override : OverrideStrategy.Fail,
+            b =>
+            {
+                if ( isOverride )
+                {
+                    b.Name = ctx.BaseOnPropertyChangedMethod!.Name;
+                }
+
+                if ( ctx.Target.IsSealed )
+                {
+                    b.Accessibility = isOverride ? ctx.BaseOnPropertyChangedMethod!.Accessibility : Accessibility.Private;
+                }
+                else
+                {
+                    b.Accessibility = isOverride ? ctx.BaseOnPropertyChangedMethod!.Accessibility : Accessibility.Protected;
+                    b.IsVirtual = !isOverride;
+                }
+            },
+            args: new { ctx } );
+
+        ctx.OnPropertyChangedMethod.Declaration = result.Declaration;
+
+        // Ensure that all required fields are generated in advance of template execution.
+        // The node selection logic mirrors that of the template's loops and conditions.
+
+        if ( ctx.OnUnmonitoredObservablePropertyChangedMethod.WillBeDefined )
+        {
+            foreach ( var node in ctx.DependencyGraph.DescendantsDepthFirst()
+                         .Where( n => n.InpcBaseHandling == InpcBaseHandling.OnUnmonitoredObservablePropertyChanged ) )
+            {
+                _ = ctx.GetOrCreateHandlerField( node );
+            }
+        }
+
+        foreach ( var node in ctx.DependencyGraph.Children
+                     .Where( node => node.InpcBaseHandling is InpcBaseHandling.OnPropertyChanged && node.HasChildren ) )
+        {
+            _ = ctx.GetOrCreateHandlerField( node );
+            _ = ctx.GetOrCreateLastValueField( node );
+        }
+    }
+
+    private static void IntroduceOnChildPropertyChangedMethod( BuildAspectContext ctx )
+    {
+        var isOverride = ctx.BaseOnChildPropertyChangedMethod != null;
```

Comment:
By default, this method should not be generated if it not invoked by the class. There should be an option (false by default) to make the code generation "future proof".

Reply:
See [the comment above](https://github.com/postsharp/Metalama.Patterns/pull/23#discussion_r1341356660).

Reply:
There will be an option.